server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java

```diff
@@ -214,25 +218,31 @@ private Translog.Operation readDocAsOp(int docIndex) throws IOException {
                 if (leaf.reader() instanceof SequentialStoredFieldsLeafReader) {
                     storedFieldsReader = ((SequentialStoredFieldsLeafReader) leaf.reader()).getSequentialStoredFieldsReader();
                     storedFieldsReaderOrd = leaf.ord;
+                    setNextSourceMetadataReader(leaf);
                 } else {
                     storedFieldsReader = null;
                     storedFieldsReaderOrd = -1;
                 }
             }
         }
+
         if (storedFieldsReader != null) {
             assert singleConsumer : "Sequential access optimization must not be enabled for multiple consumers";
             assert parallelArray.useSequentialStoredFieldsReader;
             assert storedFieldsReaderOrd == leaf.ord : storedFieldsReaderOrd + " != " + leaf.ord;
             storedFieldsReader.document(segmentDocID, fields);
         } else {
+            setNextSourceMetadataReader(leaf);
             leaf.reader().storedFields().document(segmentDocID, fields);
         }
+        final BytesReference source = fields.source() != null
+            ? addSourceMetadata(Source.fromBytes(fields.source()), segmentDocID).internalSourceRef()
+            : null;
 
         final Translog.Operation op;
         final boolean isTombstone = parallelArray.isTombStone[docIndex];
         if (isTombstone && fields.id() == null) {
-            op = new Translog.NoOp(seqNo, primaryTerm, fields.source().utf8ToString());
+            op = new Translog.NoOp(seqNo, primaryTerm, source != null ? source.utf8ToString() : "noop tombstone");
```

Comment:
Is it possible for this message to get to a user?

Reply:
Reverted